pkg/router/controller/hostindex/hostindex.go

```diff
@@ -265,7 +265,26 @@ func (r *hostRules) add(route *routev1.Route, fn RouteActivationFunc, changes *r
 	}
 
 	active, displaced := fn(changes, r.active, route)
-	r.active = active
+	// If the incoming route is a path modification to an existing route, the new
+	// route is appended to the accepted list, resulting in multiple UIDs for the
+	// same host, which isn't valid and causes undefined behavior upstream. To
+	// ensure consistency, filter the active list to ensure duplicates of the
+	// incoming route are removed (put another way, if the incoming route was
+	// accepted and active, it should replace any other routes with the same UID).
+	//
+	// The incoming route will already have been marked as a change for plugin
+	// notification, and since a route can't displace itself, there's no need to
+	// notify about the replacement. The notification of the new route state
+	// should cause an update to the existing route in the plugin without
+	// necessitating a delete/create.
+	filtered := active[:0]
+	for _, existing := range active {
+		if existing.UID == route.UID && existing.ResourceVersion != route.ResourceVersion {
+			continue
+		}
+		filtered = append(filtered, existing)
+	}
+	r.active = filtered
```

Comment:
Do you think it might be simpler to just remove the route inside `HostIndex.[aA]dd` if the paths don't match and let the add  proceed as if it was a new route?  

Aka: in here: 
https://github.com/openshift/router/blob/2b0286e8d005682382ef14f343e0903911ca7039/pkg/router/controller/hostindex/hostindex.go#L104
add something like:  
```
      hi.remove(existing, false, changes)
      newRoute = true
```

That way, if it gets re-activated back it would just be an underlying plugin route added but if it gets displaced because there is some other route holding that path, it would be a displaced route and its status would be updated to rejected as some other route holding that "host+path". 

And that would be a good test case to add: 
1. Route1 is the oldest route and holds host and path ==  www.abc.test/x/y/z  
2. Route2 is newer than Route1 and originally set to host+path == www.abc.test/foo
3. Route2 gets modified to host+path === www.abc.test/bar 
    which should let it be re-activated and admitted back in with the new/updated path.
4. Route2 gets modified to host+path == www.abc.test/x/y/z 
     which should cause Route2 to be rejected because Route1 holds that host+path.

Would that change make more sense?

Reply:
> Do you think it might be simpler to just remove the route inside `HostIndex.[aA]dd` if the paths don't match and let the add proceed as if it was a new route?

I think this is correct.  The `hi.remove` code path would remove the existing route from the inactive and active lists before the activation function ran, so the activation function would not call `changed.Displaced` on the existing route but would activate the oldest route that the existing route had been displacing.  Then the `rules.add` would re-activate the existing route iff it did not conflict with an older route.

Reply:
This is analogous to the check for host changed:
```go
	// if the host value changed, remove the old entry
	oldHost, ok := hi.routeToHost[key]
	if ok && oldHost != host {
		if existing, _, _, ok := hi.findRoute(oldHost, key); ok {
			hi.remove(existing, true, changes)
			newRoute = false
		}
	}
```
Of course we can only check for path changed after we have gotten the route using `findRoute`.  However, the cases are similar, so if using `hi.remove` and then `rules.add` is correct for the host-changed case, it makes sense to use it for the path-changed case as well.

Reply:
Great ideas, I think this works. One thing I think was missing from my mental model here is that by the time we're handling this route update, the new state is already persisted to storage, so it's safe to remove the old state immediately (because if the new state is rejected, we're not going to revert the persisted state). Awesome!

Reply:
@ramr @Miciah refactored and added new test, PTAL.